src/longsequences/longsequence.jl

```diff
@@ -84,25 +84,29 @@ end
 # Replace a LongSequence's data with a copy, copying only what's needed.
 # The user should never need to call this, as it has no outward effect on the
 # sequence.
-function orphan!(seq::LongSequence{A},
+function orphan!(seq::LongSequence,
 		 size::Integer = length(seq),
-		 force::Bool = false) where {A}
+		 force::Bool = false)
+	if !seq.shared & !force
+	    return seq
+	end
+	return _orphan!(seq, size)
+end
 
-    if !seq.shared && !force
-        return seq
-    end
+function _orphan!(seq::LongSequence{A},
+		 size::Integer = length(seq)) where {A}
 
     j, r = bitindex(seq, 1)
     data = Vector{UInt64}(undef, seq_data_len(A, size))
 
-    if !isempty(seq) && !isempty(data)
+    @inbounds if !isempty(seq) & !isempty(data)
         x = seq.data[j] >> r
         m = index(bitindex(seq, lastindex(seq))) - j + 1
         l = min(lastindex(data), m)
-        @inbounds @simd for i in 1:l-1
+        @simd for i in 1:l-1
             y = seq.data[j + i]
             data[i] = x | y << (64 - r)
-            x = y >> r
+            x = y >> (r & 63)
```

Comment:
This line was the source of the bug?

Reply:
No, actually it was just `seq.part = 1:length(seq)` instead of `seq.part = 1:size` at the end. The `& 63` just makes the bitshift compile to a single instruction. (It's needed because native bitshifts and Julia's bitshifts behave slightly different).

Reply:
To expand: In Julia, `x << 64` zeros x, whereas natively, it does nothing (since only the 6 lower bits of the bitshift integer actually does anything). That's also why I can't mask the left-shift of `y` with 63. `y` needs to be able to be zero'd.

Reply:
I'm not sure I understand:

```julia
julia> @code_native 0xFFFFFFFFFFFFFFFF >> 64
	.section	__TEXT,__text,regular,pure_instructions
; ┌ @ int.jl:448 within `>>'
; │┌ @ int.jl:448 within `<='
	xorl	%ecx, %ecx
; │└
; │ @ int.jl:448 within `>>' @ int.jl:442
	cmpq	$63, %rsi
	shrxq	%rsi, %rdi, %rdx
	cmovaq	%rcx, %rdx
; │ @ int.jl:448 within `>>'
; │┌ @ int.jl:51 within `-'
	movq	%rsi, %rax
	negq	%rax
; │└
; │┌ @ int.jl:443 within `<<'
	cmpq	$63, %rax
	shlxq	%rax, %rdi, %rax
	cmovaq	%rcx, %rax
; │└
; │┌ @ int.jl:428 within `<='
	testq	%rsi, %rsi
; │└
	cmovnsq	%rdx, %rax
	retq
; └
; ┌ @ int.jl:448 within `<invalid>'
	nopw	(%rax,%rax)
; └
```

```julia
julia> @code_native 0xFFFFFFFFFFFFFFFF >> (64 & 63)
	.section	__TEXT,__text,regular,pure_instructions
; ┌ @ int.jl:448 within `>>'
; │┌ @ int.jl:448 within `<='
	xorl	%ecx, %ecx
; │└
; │ @ int.jl:448 within `>>' @ int.jl:442
	cmpq	$63, %rsi
	shrxq	%rsi, %rdi, %rdx
	cmovaq	%rcx, %rdx
; │ @ int.jl:448 within `>>'
; │┌ @ int.jl:51 within `-'
	movq	%rsi, %rax
	negq	%rax
; │└
; │┌ @ int.jl:443 within `<<'
	cmpq	$63, %rax
	shlxq	%rax, %rdi, %rax
	cmovaq	%rcx, %rax
; │└
; │┌ @ int.jl:428 within `<='
	testq	%rsi, %rsi
; │└
	cmovnsq	%rdx, %rax
	retq
; └
; ┌ @ int.jl:448 within `<invalid>'
	nopw	(%rax,%rax)
; └
```

They look the same to me.

Reply:
You just need to put it in a function:
```
julia> f(a, b) = >>>(a, b & 63)
f (generic function with 1 method)

julia> code_native(>>>, (Int, Int), debuginfo=:none)
    .section    __TEXT,__text,regular,pure_instructions
    xorl    %ecx, %ecx
    cmpq    $63, %rsi
    shrxq   %rsi, %rdi, %rdx
    cmovaq  %rcx, %rdx
    movq    %rsi, %rax
    negq    %rax
    cmpq    $63, %rax
    shlxq   %rax, %rdi, %rax
    cmovaq  %rcx, %rax
    testq   %rsi, %rsi
    cmovnsq %rdx, %rax
    retq
    nopw    (%rax,%rax)

julia> code_native(f, (Int, Int), debuginfo=:none)
    .section    __TEXT,__text,regular,pure_instructions
    shrxq   %rsi, %rdi, %rax
    retq
    nopw    %cs:(%rax,%rax)
```

Reply:
It guess it would be slightly faster still to change `data[i] = x | y << (64 - r)` to `data[i] = x | y << unsigned(64 - r)`

Reply:
Oh yep that works for me too. So julia is doing some extra stuff to zero a variable, whereas shrxq only cares about the 6 LSBs? I should probably see if I can use this trick in some of my bitpar counting stuff.

Reply:
@BenJWard Yes, in Julia's it's equivalent to this like this:
```
function >>>(n::Signed, s::Signed)
    if s < 0
        return n << -s
    elseif s > 63
        return zero(n)
    else
        return LLVM_RIGHT_SHIFT(n, s)
    end
end
```
Using an `unsigned` for `s` removes the first branch, but not the second. Using `& 63` removes both.